### PR TITLE
CSPL- 1254: Added DownloadApp API for AWS and MINIO client

### PR DESCRIPTION
**Pull request by gaurav-splunk**
This PR covers the following changes - 
1. Introduces **DownloadApp()** API for both AWS and MINIO clients.
2. Covers UT for both clients.
3. Added some other framework to cover download APIs in UTs.
3. It also has some code changes to satisfy go static analysis checks(such as using lower case for error strings), hence the changes in all other files.

### CODE REVIEW

File: pkg/splunk/enterprise/util.go [hunk 1]
```diff
@@ -506,6 +506,21 @@ func (s3mgr *S3ClientManager) GetAppsList() (splclient.S3Response, error) {
...
+	c, err := s3mgr.getS3Client(s3mgr.client, s3mgr.cr, s3mgr.appFrameworkRef, s3mgr.vol, s3mgr.location, s3mgr.initFn)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.Client.DownloadApp(remoteFile, localFile)
+	if err != nil {
+		return err
+	}
+	return nil
```
Comment (sgontla) [inline pkg/splunk/enterprise/util.go:521]: nit: `return err` ?

File: pkg/splunk/test/minioclient.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	return objectCh
+}
+
+// FGetObject is a mock call to download file/app from minio client.
+// It just does some error checking.
+func (mockClient MockMinioS3Client) FGetObject(ctx context.Context, bucketName string, remoteFileName string, localFileName string, opts minio.GetObjectOptions) error {
+
+	var err error
+	if remoteFileName == "" || localFileName == "" {
+		err = fmt.Errorf("empty remoteFileName/localFileName")
```
Comment (sgontla) [inline pkg/splunk/test/minioclient.go:97]: Nit: You may want to make it explicit about the error / print the values?

File: pkg/splunk/client/awss3client.go [hunk 3]
```diff
@@ -195,6 +205,34 @@ func (awsclient *AWSS3Client) GetAppsList() (S3Response, error) {
 	return s3Resp, nil
 }
 
+// DownloadApp downloads the app from remote storage to local file system
+func (awsclient *AWSS3Client) DownloadApp(remoteFile, localFile string) (bool, error) {
+	scopedLog := log.WithName("DownloadApp")
```
Comment (sgontla) [inline pkg/splunk/client/awss3client.go:210]: here also.

File: pkg/splunk/client/awss3client.go [hunk 4]
```diff
@@ -195,6 +205,34 @@ func (awsclient *AWSS3Client) GetAppsList() (S3Response, error) {
...
+
+	var numBytes int64
+	file, err := os.Create(localFile)
+	if err != nil {
+		scopedLog.Error(err, "Unable to open file", "localFile", localFile)
```
Comment (jryb) [inline pkg/splunk/client/awss3client.go:215]: nit: should this be lower case too?

File: pkg/splunk/client/minioclient_test.go [hunk 5]
```diff
@@ -67,3 +72,434 @@ func TestGetMinioInitContainerCmd(t *testing.T) {
 		t.Errorf("Got incorrect Init container cmd")
 	}
 }
+
+func TestMinioGetAppsListShouldNotFail(t *testing.T) {
```
Comment (jryb) [inline pkg/splunk/client/minioclient_test.go:76]: No code comment, just wanted to say thanks for writing these. 🥇

**Review by jryb — APPROVED** [review]

File: pkg/splunk/client/awss3client.go (hunk 4 above)
Comment (smohan-splunk) [reply pkg/splunk/client/awss3client.go:215]: May be a re-run of the static analysis is needed to detect these

**Review by smohan-splunk — APPROVED** [review]

**Comment by smohan-splunk** [comment]
@jryb, @gaurav-splunk  Just checking, don't we need a minio equivalent of functions such as `TestClusterMasterGetAppsListForAWSS3ClientShouldNotFail`?

File: pkg/splunk/client/minioclient.go [hunk 6]
```diff
@@ -138,6 +140,29 @@ func (client *MinioClient) GetAppsList() (S3Response, error) {
 	return s3Resp, nil
 }
 
+// DownloadApp downloads an app package from remote storage
+func (client *MinioClient) DownloadApp(remoteFile string, localFile string) (bool, error) {
+	scopedLog := log.WithName("DownloadApp")
+
+	file, err := os.Create(localFile)
+	if err != nil {
```
Comment (smohan-splunk) [inline pkg/splunk/client/minioclient.go:148]: Will this fail if a file already exists?

File: pkg/splunk/client/awss3client.go [hunk 7]
```diff
@@ -195,6 +205,34 @@ func (awsclient *AWSS3Client) GetAppsList() (S3Response, error) {
...
+		})
+	if err != nil {
+		scopedLog.Error(err, "Unable to download item %s, %v", remoteFile, err)
+		return false, err
+	}
+
+	scopedLog.Info("File Downloaded", "remoteFile: ", remoteFile, "numBytes: ", numBytes, "localFile: ", localFile)
+
+	return true, err
+}
```
[the remark this answers is not in the document]
Comment (gaurav-splunk) [reply pkg/splunk/client/awss3client.go:234]: Yeah seems a fair point, I will look for options in the download api to match the tag

File: pkg/splunk/client/awss3client.go (hunk 4 above)
Comment (gaurav-splunk) [reply pkg/splunk/client/awss3client.go:215]: I think it only checks for error messages in `fmt` and not the custom error strings.

File: pkg/splunk/client/minioclient.go (hunk 6 above)
Comment (gaurav-splunk) [reply pkg/splunk/client/minioclient.go:148]: Nope, it does not

**Comment by gaurav-splunk** [comment]
> @jryb, @gaurav-splunk Just checking, don't we need a minio equivalent of functions such as `TestClusterMasterGetAppsListForAWSS3ClientShouldNotFail`?

We can, but it won't improve the code coverage and I thought it would be one too many redundant tests.

**Review by sgontla — APPROVED** [review]

**Review by smohan-splunk — APPROVED** [review]

**Pull request merged**